sympy/parsing/latex/lark/grammar/latex.lark

```diff
@@ -104,38 +104,14 @@ GTE: "\\geq" | "\\ge" | "\\geqslant"
 DIV_SYMBOL: CMD_DIV | DIV
 MUL_SYMBOL: MUL | CMD_TIMES | CMD_CDOT
 
-// Greek symbols
-// TODO: Shouold we include the uppercase variants for the symbols where the uppercase variant doesn't have a separate meaning?
-ALPHA: "\\alpha"
-BETA: "\\beta"
-GAMMA: "\\gamma"
-DELTA: "\\delta" // TODO: Should this be included? Delta usually denotes other things.
-EPSILON: "\\epsilon" |  "\\varepsilon"
-ZETA: "\\zeta"
-ETA: "\\eta"
-THETA: "\\theta" | "\\vartheta"
-// TODO: Should I add iota to the list?
-KAPPA: "\\kappa"
-LAMBDA: "\\lambda" // TODO: What about the uppercase variant?
-MU: "\\mu"
-NU: "\\nu"
-XI: "\\xi"
-// TODO: Should there be a separate note for transforming \pi into sympy.pi?
-RHO: "\\rho" | "\\varrho"
-// TODO: What should we do about sigma?
-TAU: "\\tau"
-UPSILON: "\\upsilon"
-PHI: "\\phi" | "\\varphi"
-CHI: "\\chi"
-PSI: "\\psi"
-OMEGA: "\\omega"
-
-GREEK_SYMBOL: ALPHA | BETA | GAMMA | DELTA | EPSILON | ZETA | ETA | THETA | KAPPA
-    | LAMBDA | MU | NU | XI | RHO | TAU | UPSILON | PHI | CHI | PSI | OMEGA
-
-DIFFERENTIAL: "d" | "\\text{d}" | "\\mathrm{d}"
-
-SYMBOL: /[a-zA-Z]/
+%import .greek_symbols.GREEK_SYMBOL
+
+UPRIGHT_DIFFERENTIAL_SYMBOL: "\\text{d}" | "\\mathrm{d}"
+LOWERCASE_LETTER_D: "d"
```

Comment:
Keep in mind that extracting "d" from SYMBOL makes the code a lot more complicated. This project is meant to be continued in the future... we should try to find a way to do this without adding ad-hoc hooks for special meanings of expressions.

Reply:
> makes the code a lot more complicated

I think the multiple pass idea makes the code even more complicated. I looked through the docstring for the Mathematica parser and it looks like it does 4 passes.

Also, I don't think looking at how the Mathematica parser handles things is not a fruitful idea because that is a CAS-to-CAS parser. A LaTeX to SymPy parser is a typesetting language to CAS parser, which is an inherently different beast, as I've mentioned multiple times.

Reply:
If we want simplicity, then I think we should go the route @sylee957 mentioned, and only support `\text{d}` and `\mathrm{d}` as differentials because of the ISO standard.

Reply:
> I think the multiple pass idea makes the code even more complicated.

It is better to have multiple simple passes than one complicated one especially if it avoids having hard-coded rules and if separating the passes also separates concerns. Any decision you make about how to handle `d` here will be disliked by some users who will then want a way to control it. Hard-coded rules at the parse level make it more difficult to customise the behaviour in future.

It is better to have a first pass that handles parsing only and uses unambiguous parsing rules corresponding to the LaTeX formal grammar (which is not ambiguous), with the output being effectively a concrete or abstract syntax tree. This first step should not make any decisions that any users might reasonably complain about like disallowing `d` as a variable name or deciding that `dx` is a product or a differential. Then any steps after that that are ambiguous or not always desired should be *optionally* handled in the subsequent passes. These subsequent steps are much easier to implement and/or customise as tree-to-tree operations than when built into the parser.

Having a single pass might seem simpler now but it will make things more complicated in future for anyone who might try to extend or improve this parser. The difficulty you have right now in handling `d` demonstrates precisely the difficulties that future contributors will have when trying to build on this work to add or customise other features.

Reply:
> It is better to have a first pass that handles parsing only and uses unambiguous parsing rules corresponding to the LaTeX formal grammar (which is not ambiguous), with the output being effectively a concrete or abstract syntax tree.

Unless I am mistaken, what you're suggesting sounds like a complete revamp of the project. I'm not sure how feasiable this is, given that my GSoC project deadline is 6 days away. This is something we should have discussed and decided at the start. I am more and more inclined to take @sylee957's suggestion and only allow `\text{d}` and `\mathrm{d}` as differentials.

> LaTeX formal grammar (which is not ambiguous)

It doesn't matter if the (La)TeX formal grammar is ambiguous or not. We are, in essence, trying to extract mathematical meaning from a language meant for typesetting. Ambiguity is inevitable. TeX doesn't care what `d` is. All it cares about is whether to italicize it not. For TeX, the decision is simple: default to italicize, unless told otherwise.

We have the following ambiguities so far:
* Do we parse `d` as a differential or as a variable?
* Do we parse `e` and `\pi` as the constants or are arbitrary variables and function names?
* Do we parse `f(x)` or, more generally, `u(t)` as function applications or as a multiplication of two variables?

All of these decisions, ultimately, depend on the user. We as library writers cannot make these decisions, and as far as I understand, each of these require slightly different grammars. Once again, because TeX only cares about looks (i.e., visual output), it doesn't care about the mathematical meanings and is unfazed by these ambiguities. We cannot extract the user's intent from these LaTeX strings.

What we _can_ do, however, as an initial measure, is to have a default which satisfies most usecases. Non-standard or niche uses comes after that. What I have currently prevents the user from writing something like `dt` as a multiplication. That's it. Everything else works fine.

> The difficulty you have right now in handling d demonstrates precisely the difficulties that future contributors will have when trying to build on this work to add or customise other features.

I can make the behavior I described above happen even without adding another rule, if that's what you care about. I thought that doing it this way conveys the special nature of `d` in particular, and my intent as a programmer, better. Plus, as @sylee957 mentioned, it's best to have as much brancing logic in the grammar as possible instead of in the transformer class. That's why I broke up the `relation` rule in the grammar, even though one could argue that it complicated the grammar.

Reply:
> I can make the behavior I described above happen even without adding another rule, if that's what you care about. I thought that doing it this way conveys the special nature of `d` in particular, and my intent as a programmer, better.

I've gone ahead and removed the special casing. Everything still works the same way as before.

Reply:
To be clear I am not objecting to the changes here but just commenting on the design decisions for future consideration.

I realise that it would be difficult to change these things at this late stage in the project but when making any changes now we should still be aware of whether or not those changes now are compatible with how we might want to change or extend the code in future.

> All of these decisions, ultimately, depend on the user. We as library writers cannot make these decisions,

Agreed.

> and as far as I understand, each of these require slightly different grammars.

No, this is not true. They only seem to require different "grammars" because you are mixing up grammar with other things in the parse phase. LaTeX has an unambiguous grammar and the best way to parse it is that the first pass parses that grammar in an unambiguous way. A parser that matches the actual grammar perfectly is a solid base on top of which any other LaTeX parsing functionality can be built. If the base parsing phase already makes any undesirable decisions then it is impossible to undo those in any code that builds on top.

Reply:
>  LaTeX has an unambiguous grammar and the best way to parse it is that the first pass parses that grammar in an unambiguous way. A parser that matches the actual grammar perfectly is a solid base on top of which any other LaTeX parsing functionality can be built.

The only thing that can be very reusable and unambiguous, about LaTeX are tokens
For example, we can add everything in 
https://katex.org/docs/supported
without much technical difficulty.

However, I'm not very concluded about how to make the second phase (the CFG part) more reusable,
if it is very much thin, it won't be much different than the tokenizer itself.
And if we just attempt to parse every mathematical grammar users may need, it can likely be dead in ambiguity and complexity.
So at this stagge, I think that it is just important to focus on clarity of implementation, even if it is small, than having more rich set of features.

Reply:
I would further argue that in the future we could add a Machine Learning stage after the transformer that applies advanced transformations based on statistical learning of real world use cases. Maybe we shouldn't customize the Lark parser too much.